service/discovery/aws/storage.go

```diff
@@ -0,0 +1,382 @@
+				URL:                 bucket.endpoint,
+				TransportEncryption: &encryptionAtTransmit,
+			},
+		})
+	}
+	return
+}
+func (b bucket) String() string {
+	return fmt.Sprintf("[ARN: %v, Name: %v, Creation Time: %v, Number of objects: %v]", b.arn, b.name, b.creationTime, b.numberOfObjects)
+}
+
+// NewAwsStorageDiscovery constructs a new awsS3Discovery initializing the s3-client and isDiscovering with true
+func NewAwsStorageDiscovery(cfg aws.Config) discovery.Discoverer {
+	return &awsS3Discovery{
+		client:        s3.NewFromConfig(cfg),
+		buckets:       nil,
+		isDiscovering: true,
+	}
+}
+
+// S3ListBucketsAPI is the interface for the List function (used for mock testing)
+type S3ListBucketsAPI interface {
+	ListBuckets(ctx context.Context,
+		params *s3.ListBucketsInput,
+		optFns ...func(*s3.Options)) (*s3.ListBucketsOutput, error)
+}
+
+// getBuckets returns all buckets
+func (d *awsS3Discovery) getBuckets() (err error) {
+	log.Println("Getting buckets in s3...")
+	var resp *s3.ListBucketsOutput
+	resp, err = d.client.ListBuckets(context.TODO(), &s3.ListBucketsInput{})
+	if err != nil {
+		log.Error("Error occurred.")
+		var oe *smithy.OperationError
+		if errors.As(err, &oe) {
+			log.Printf("failed to call service: %s, operation: %s, error: %v", oe.Service(), oe.Operation(), oe.Unwrap())
+		}
+		return
+	}
+	var region string
+	log.Printf("Retrieved %v buckets.", len(resp.Buckets))
+	for _, b := range resp.Buckets {
+		//d.buckets = resp
+		region, err = d.getRegion(aws.ToString(b.Name))
+		if err != nil {
+			return
+		}
+		d.buckets = append(d.buckets, bucket{
+			arn:          "arn:aws:s3:::" + *b.Name,
+			name:         aws.ToString(b.Name),
+			creationTime: aws.ToTime(b.CreationDate),
+			region:       region,
+			endpoint:     "https://" + aws.ToString(b.Name) + ".s3." + region + ".amazonaws.com",
+			// ToDo: Implement method for retrieving the number of objects per bucket (if needed)
+			numberOfObjects: -1,
+		})
+	}
+	return
+}
+
+// getEncryptionAtRest gets the bucket's encryption configuration
+func (d *awsS3Discovery) getEncryptionAtRest(bucket string) (e voc.AtRestEncryption, err error) {
+	log.Printf("Checking encryption for bucket %v.\n", bucket)
+	input := s3.GetBucketEncryptionInput{
+		Bucket:              aws.String(bucket),
+		ExpectedBucketOwner: nil,
+	}
+	var resp *s3.GetBucketEncryptionOutput
+
+	resp, err = d.client.GetBucketEncryption(context.TODO(), &input)
+	if err != nil {
+		log.Error("Error occurred")
+		var oe *smithy.OperationError
+		if errors.As(err, &oe) {
+			log.Printf("failed to call service: %s, operation: %s, error: %v", oe.Service(), oe.Operation(), oe.Unwrap())
+		}
+		return
+	}
+	log.Println("Bucket is encrypted.")
+	e.Algorithm = string(resp.ServerSideEncryptionConfiguration.Rules[0].ApplyServerSideEncryptionByDefault.SSEAlgorithm)
+	if e.Algorithm == string(types.ServerSideEncryptionAes256) {
+		e.KeyManager = "SSE-S3"
+	} else {
+		e.KeyManager = "SSE-KMS"
+	}
+	e.Enabled = true
```

Comment:
it's always encrypted?

Reply:
The way the AWS API call works: Either encryption is in place and returned in the response field; or there is no encryption and the API call returns an error. In the case of error, the struct field will be kept as initialized with its zero values (==false). But I can add a comment to make that more clear.

But as I think about it, we also have to distinguish then between errors that are API Call failures (which should be propagated to the caller) and errors that are just no-encryption-is-set-errors. I have to look into that and maybe handle it via a switch statement.